source/developers/sts-for-operator.rst

```diff
@@ -0,0 +1,132 @@
+.. _minio-sts-operator:
+
+===============================================
+Security Token Service (STS) for MinIO Operator
+===============================================
+
+.. default-domain:: minio
+
+.. contents:: Table of Contents
+   :local:
+   :depth: 2
+
+.. versionadded:: Operator v5.0.0
+
+   The MinIO Operator supports a set of API calls that allows an application to obtain STS credentials for a MinIO Tenant.
```

Comment:
Q: @pjuarezd what problem is this solving in K8s vs sending the API request to the tenant itself?

Also, does the `AssumeRoleWithWebIdentity` API remain the same? How does the application specify the tenant it wants to auth against?

Reply:
this is a excelent question a bit longer to answer, sorry for the long text.

`Operator STS API` solves some problems:

* Authenticate applications in kubernetes to access MinIO using a kubernetes native authentication mechanism: using kubernetes Service Accounts Tokens, the `Service Account` is a core concept of the Role-based Access Control (RBACK) Native kubernetes Authenticatio and authorization.

* Allows `Infrastructure as code` like configuration to access MinIO in kubernetes, by using  the Tenant CRD + PolicyBinding CRD.

* Allows a "credential-less" authentication in the client code, no need to create and specify IAM user to gain access to MinIO API, just instantiate the SDK client and get access to the 

We had often been asked to provide a way to get credentials for an application in a "credentialless" way, similar to how AWS IMDS allows to get the STS Token of the instance profile assigned to the EC2 instance. [AWS doc example here](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html#instance-metadata-security-credentials)

We also have been asked to provide an mechanism to authenticate using kubernetes Service accounts similar to [AWS IRSA](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html), Operator STS does it.

Reply:
Thanks @pjuarezd - so that comes back to how to use the API itself. From what I'm reading in a previous comment, the application would need to have access to the JWT from Kubernetes itself - so that JWT has to be passed to the applications somehow and then maintained in case it ever changes.

that still leaves a question of how the Operator knows which Tenant to generate credentials for. Is it purely by the service account token being namespaced already?

Reply:
yeah, regarding that second question

> Also, does the AssumeRoleWithWebIdentity API remain the same? How does the application specify the tenant it wants to auth against?

The request is not the same that doing it directly to the tenant, as the tenant knows nothing about kuberneters RBAC, Operator will do that authentication, and then request an STS token to the tenant using root credentials, somenthing like this:

![Authentication with service Accoun](https://user-images.githubusercontent.com/1334362/236016701-fc871cbc-d247-4b06-b648-6df2a1d7b0d4.jpg)

Reply:
That makes sense. I guess here then the one requirement is we have to assume the K8s admin has a way of passing that SA JWT to the application and keeping it up-to-date, otherwise things crap out. @djwfyi I would suggest mentioning something to that effect if it's not there already and I just dumb brained past it.

Reply:
The bridge to know which tenant the sdk App is getting credentials for is the PolicyBinding Custom resource.

1) The client specify the namespace of the tenant in the STS endpoint, ie using the AWS SDK code would look like this

```python
stsUrl = urlparse(f"https://sts.minio.operator.svc.cluster.local:4223/tenant-1")
sts = boto3.client('sts', endpoint_url=stsUrl.geturl(), verify=False)
assumed_role_object = sts.assume_role_with_web_identity(
    RoleArn="arn:aws:iam::111111111:dummyroot",
    RoleSessionName="session 1",
    WebIdentityToken=sa_jwt
)

s3_client = boto3.resource('s3',
    aws_access_key_id=credentials['AccessKeyId'],
    aws_secret_access_key=credentials['SecretAccessKey'],
    aws_session_token=credentials['SessionToken'],
    endpoint_url=tenantUrl.geturl(), verify=False)
```
2) a `PolicyBinding` object have to exist in the tenant namespace, and specify service account name of the JWT, and a list of IAM policies to authorize.


```yaml
apiVersion: sts.min.io/v1alpha1
kind: PolicyBinding
metadata:
  name: binding-1
  namespace: tenant-1
spec:
  application:
    namespace: application-namespace
    serviceaccount: aplication-service-account
  policies:
    - bucket1-rw
    - bucket2-read
```


Because the PolicyBinding is a Custom resource and can be specified as yaml, this opens the door to Infrastructure as code authentication, managed by the kubernetes manager. I am thinking of ArgoCD maybe?

Reply:
> That makes sense. I guess here then the one requirement is we have to assume the K8s admin has a way of passing that SA JWT to the application and keeping it up-to-date, otherwise things crap out. @djwfyi I would suggest mentioning something to that effect if it's not there already and I just dumb brained past it.

we not necesarelly need to assume the k8s admin has a way of passing the SA JWT, because the application running in a pod in kubernetes is the one that will have to pass the JWT to the SDK's, for that, the pods always have the JWT   mounted in the filesystem  in the path `/var/run/secrets/kubernetes.io/serviceaccount/token`

I did a bit of explanation of where the JWT comes from in this comment https://github.com/minio/docs/pull/841#discussion_r1183996140

Reply:
STS are temporary credentials, that expire and need to be recreated, SDK's currently have a verification flow, if the STS token is expired, then sdk library "should" requests  a new STS token before the next API call, but that is up to how the SDK client is implemented.

SA JWT also has expiration, yes that token expiration is a whole topic, there is long-lived tokens and short-lived tokens, automatic mounted tokens, manually created, etc. But we could just limit ourselfs to require a `valid, non expired Service account Token` regardless if is the one mounted in the pod, or another requested using K8s API by the SDK application.

This medium article might do a better explaining to Service account JWT that me https://medium.com/@th3b3ginn3r/understanding-service-accounts-in-kubernetes-e9d2abe19df8

Reply:
This is all great info. Added applicable changes throughout the PR.